File: python/程序员面试金典/面试题 01.08. 零矩阵.py
from typing import List
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m,n=len(matrix),len(matrix[0])
        row = set()
        col = set()
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    row.add(i)
                    col.add(j)
        for i in range(m):
            for j in range(n):
                if i in row or j in col:
                    matrix[i][j] = 0

# 空间复杂度O(1)
# 先记录第一行第一列是否有0，然后利用第一行第一列作为标记，先清除非第一行或非第一列的数据，最后根据第一列第一列是否有0来清除第一行第一列
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        m,n=len(matrix),len(matrix[0])
        isFirstRowHaveZero = False
        isFirstColHaveZero = False
        for i in range(m):
            if matrix[i][0] == 0:
                isFirstColHaveZero = True

        for i in range(n):
            if matrix[0][i] == 0:
                isFirstRowHaveZero = True

        for i in range(1,m):
            for j in range(1,n):
                if matrix[i][j] == 0:
                    matrix[0][j] = 0
                    matrix[i][0] = 0

        for i in range(1,m):
            for j in range(1,n):
                if matrix[0][j] == 0 or matrix[i][0] ==0:
                    matrix[i][j] = 0

        for i in range(m):
            if isFirstColHaveZero:
                matrix[i][0] = 0

        for i in range(n):
            if isFirstRowHaveZero:
                matrix[0][i] = 0

File: go/程序员面试金典/面试题 01.08. 零矩阵.go
package main

func setZeroes(matrix [][]int) {
	m, n := len(matrix), len(matrix[0])
	isRowZero, isColZero := false, false
	for i := 0; i < m; i++ {
		if matrix[i][0] == 0 {
			isColZero = true
		}
	}

	for i := 0; i < n; i++ {
		if matrix[0][i] == 0 {
			isRowZero = true
		}
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if matrix[i][j] == 0 {
				matrix[i][0] = 0
				matrix[0][j] = 0
			}
		}
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if matrix[i][0] == 0 || matrix[0][j] == 0 {
				matrix[i][j] = 0
			}
		}
	}

	for i := 0; i < m; i++ {
		if isColZero {
			matrix[i][0] = 0
		}
	}

	for i := 0; i < n; i++ {
		if isRowZero {
			matrix[0][i] = 0
		}
	}
}
